### PR TITLE
[AIR] Change the name of `GBDTTrainable` dynamically

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Mirrors the logic already present in BaseTrainer. Makes for nicer UX & logging. Works around a core bug/oversight where `__repr__` has to be directly implemented in the class (inherited `__repr__` is ignored) - bug tracked [here](https://github.com/ray-project/ray/issues/25817).

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/gbdt_trainer.py [hunk 1]
```diff
@@ -224,4 +224,6 @@ def default_resource_request(cls, config):
                     scaling_config_dataclass, ray_params_cls, default_ray_params
                 ).get_tune_resources()
 
+        GBDTTrainable.__name__ = trainer_cls.__name__
```
Comment (richardliaw) [inline python/ray/train/gbdt_trainer.py:227]: this is a bit hacky; can we instead have it render in the __repr__ of the class only?

Comment (Yard1) [reply python/ray/train/gbdt_trainer.py:227]: This is what we are doing in DataParallelTrainer as well (and indeed, what happens inside `wrap_trainable` - and this class inherits from the return of that). I honestly think it's fine

Comment (Yard1) [reply python/ray/train/gbdt_trainer.py:227]: actually, sure, let me change that

Comment (Yard1) [reply python/ray/train/gbdt_trainer.py:227]: ok so this is odd but `__repr__` is already set correctly (as it inherits from `wrap_function`). Not sure why it's not showing up in logs as intended.

**Comment by richardliaw** [comment]
Can you add a test here?

**Comment by Yard1** [comment]
@richardliaw added!

**Review by richardliaw — APPROVED** [review]

**Pull request merged**